constraints/miden-vm/memory.air

```diff
@@ -0,0 +1,64 @@
+def MemoryAir
+
+trace_columns:
+    main: [s[2], ctx, addr, clk, v[4], d[2], t, dummy]
+
+public_inputs:
+    stack_inputs: [16]
+
+random_values:
+    alphas: [9]
+
+boundary_constraints:
+    # This is a dummy trace column to satisfy requirement of at least one boundary constraint.
+    enf dummy.first = 0
+
+integrity_constraints:
+    let delta_c = ctx' - ctx
+    let delta_a = addr' - addr
+    let delta_i = clk' - clk - 1
+
+    # n_0 = 1 when context changes and 0 otherwise.
+    let n_0 = delta_c * t'
+    # if context remains the same, n_1 = 1 when address changes and 0 otherwise.
+    let n_1 = delta_a * t'
+
+    # Enforce that n_0 must be binary.
+    enf n_0^2 - n_0 = 0
+    # Enforce that when context changes, n_0 = 1
+    enf (1 - n_0) * delta_c = 0
+    # Enforce that n_1 must be binary. An additional multiplier ensures that the check of n_1 
+    # occurs only if the context does not change.
+    enf (1 - n_0) * (n_1^2 - n_1) = 0
+    # Enforce that if context remains the same, n_1 = 1 when address changes and 0 otherwise.
+    enf (1 - n_0) * (1 - n_1) * delta_a = 0
+
+    # Enforce that values in the selectior columns must be binary.
+    # s[0] is set to 0 for write operations and to 1 for read operations
+    enf s[i]^2 - s[i] = 0 for i in (0..2)
+
+    # Enforce that s[1] = 1 when the operation is a read and `ctx` and `addr` columns are both 
+    # unchanged.
+    enf (1 - n_0) * (1 - n_1) * s[0]' * (1 - s[1]') = 0
+    # Enforce that when either the context changed, the address changed, or the operation is a 
+    # write, then s[1] = 0
+    enf (n_0 + (1 - n_0) * n_1 + (1 - s[0]')) * s[1]' = 0
+
+    let d_next_aggr = 2^16 * d[1]' + d[0]'
+
+    # Enforce that values of context (`ctx`), address (`addr`), and clock cycle (`clk`) grow 
+    # monotonically
+    enf (n_0 * delta_c + (1 - n_0) * (n_1 * delta_a + (1 - n_1) * delta_i)) - d_next_aggr = 0
```

Comment:
If we are planning to support selector syntax for v0.3, then this could be something like:
```
match enf:
  d_next_agg = delta_c when n_0
  d_next_agg = delta_a when (1 - n_0) * n_1
  d_next_agg = delta_i when (1 - n_0) * (1 - n_1)
```
A couple other thoughts regarding improving readability:
- Instead of `n_0` and `n_1` we could probably use `n0` and `n1`, or maybe even a vector `n` with two items - though, not sure we need to go as far as using a vector.
- I wonder if we should allow syntactic sugar for logical expressions on the right side of `when`.

For example, the above could look like:
```
match enf:
  d_next_agg = delta_c when n0
  d_next_agg = delta_a when !n0 & n1
  d_next_agg = delta_i when !n0 & !n1
```